types/validation_test.go

```diff
@@ -0,0 +1,213 @@
+				},
+			},
+			state:   validState,
+			wantErr: true,
+			errMsg:  "b version mismatch",
+			isFraud: false,
+		},
+		{
+			name: "Invalid height",
+			block: &Block{
+				Header: Header{
+					Version:         validBlock.Header.Version,
+					Height:          11,
+					Time:            uint64(currentTime.UnixNano()),
+					AppHash:         [32]byte{1, 2, 3},
+					LastResultsHash: [32]byte{4, 5, 6},
+					ProposerAddress: []byte("proposer"),
+					DataHash:        [32]byte(GetDataHash(validBlock)),
+				},
+			},
+			state:   validState,
+			wantErr: true,
+			errMsg:  "height mismatch",
+			isFraud: true,
+		},
+		{
+			name: "Invalid AppHash",
+			block: &Block{
+				Header: Header{
+					Version:         validBlock.Header.Version,
+					Height:          10,
+					Time:            uint64(currentTime.UnixNano()),
+					AppHash:         [32]byte{9, 9, 9},
+					LastResultsHash: [32]byte{4, 5, 6},
+					ProposerAddress: []byte("proposer"),
+					DataHash:        [32]byte(GetDataHash(validBlock)),
+				},
+			},
+			state:   validState,
+			wantErr: true,
+			errMsg:  "AppHash mismatch",
+			isFraud: true,
+		},
+		{
+			name: "Invalid LastResultsHash",
+			block: &Block{
+				Header: Header{
+					Version:         validBlock.Header.Version,
+					Height:          10,
+					Time:            uint64(currentTime.UnixNano()),
+					AppHash:         [32]byte{1, 2, 3},
+					LastResultsHash: [32]byte{9, 9, 9},
+					ProposerAddress: []byte("proposer"),
+					DataHash:        [32]byte(GetDataHash(validBlock)),
+				},
+			},
+			state:   validState,
+			wantErr: true,
+			errMsg:  "LastResultsHash mismatch",
+			isFraud: true,
+		},
+		{
+			name: "Future block time",
+			block: &Block{
+				Header: Header{
+					Version:         validBlock.Header.Version,
+					Height:          10,
+					Time:            uint64(currentTime.Add(2 * MaxDrift).UnixNano()),
+					AppHash:         [32]byte{1, 2, 3},
+					LastResultsHash: [32]byte{4, 5, 6},
+					ProposerAddress: []byte("proposer"),
+				},
+			},
+			state:   validState,
+			wantErr: true,
+			errMsg:  "Sequencer",
+			isFraud: true,
+		},
+		{
+			name: "Invalid proposer address",
+			block: &Block{
+				Header: Header{
+					Version:         validBlock.Header.Version,
+					Height:          10,
+					Time:            uint64(currentTime.UnixNano()),
+					AppHash:         [32]byte{1, 2, 3},
+					LastResultsHash: [32]byte{4, 5, 6},
+					ProposerAddress: []byte{},
+				},
+			},
+			state:   validState,
+			wantErr: true,
+			errMsg:  "no proposer address",
+			isFraud: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := tt.block.ValidateWithState(tt.state)
+			if tt.wantErr {
+				assert.Error(t, err)
+				assert.Contains(t, err.Error(), tt.errMsg)
+				if tt.isFraud {
+					require.True(t, errors.Is(err, fraud.ErrFraud))
+				} else {
+					require.False(t, errors.Is(err, fraud.ErrFraud))
+				}
+			} else {
+				assert.NoError(t, err)
```

Comment:
Instead of testing using string compare you should remove errMsg and just use errors.Is with the expected type of fraud